### PR TITLE
Add MPAS-analysis v1

### PR DESCRIPTION
**Pull request by milenaveneziani**
This update does two main things:
1) Update to MPAS-Analysis release version v1.0;
2) Adds a python script that downloads the input data necessary to run a-prime from the lcrc public html repository (these input data includes masking files for both the atmosphere and the ocn/ice parts, region mask files for MPAS, and observation files for all components).

### CODE REVIEW

**Comment by milenaveneziani** [comment]
**Things to do before merging**:

- [ ] Download data and test on blues/anvil
- [ ] Download data and test on OLCF
- [ ] Download data and test on theta
- [ ] Download data and test on acme1/aims4
- [ ] Update to official MPAS-Analysis v1 version (tag)

**Comment by milenaveneziani** [comment]
Testing (including fresh download of all input data) was successful on edison.

File: run_aprime.bash [hunk 1]
```diff
@@ -310,31 +317,32 @@ export plots_dir=$plots_base_dir/$plots_dir_name
...
+export remap_files_dir=$projdir/diagnostics/a-prime/maps
+export GPCP_regrid_wgt_file=$projdir/diagnostics/a-prime/maps/$test_atm_res-to-GPCP.conservative.wgts.nc
+export CERES_EBAF_regrid_wgt_file=$projdir/diagnostics/a-prime/maps/$test_atm_res-to-CERES-EBAF.conservative.wgts.nc
+export ERS_regrid_wgt_file=$projdir/diagnostics/a-prime/maps/$test_atm_res-to-ERS.conservative.wgts.nc
+
+# Set ocn/ice specific paths to region mask and mapping files, and to observation files
+export mpas_mappingDirectory=$projdir/diagnostics/mpas_analysis/maps
+export mpas_regionMaskDirectory=$projdir/diagnostics/mpas_analysis/region_masks
+export obs_ocndir=$projdir/diagnostics/observations/Ocean
+export obs_seaicedir=$projdir/diagnostics/observations/SeaIce
```
Comment (milenaveneziani) [inline run_aprime.bash]: the piece below with commented out lines will go away before merging.

File: run_aprime.bash [hunk 2]
```diff
@@ -373,35 +381,43 @@ fi
...
+  conda activate e3sm_unified_1.2.0_py2.7_nox
   export NCO_PATH_OVERRIDE=No
 elif [ $machname == "lanl" ]; then
   module unload python
   module use $projdir/modulefiles/all
-  module load e3sm-unified/1.1.3
+  module load e3sm-unified/1.2.0
 elif [ $machname == "anvil" ]; then
-  unset LD_LIBRARY_PATH
-  soft add +e3sm-unified-1.1.3-nox
```
Comment (milenaveneziani) [inline run_aprime.bash]: same with lines below. Just wanted to test on anvil first before deleting.

File: run_aprime.bash [hunk 3]
```diff
@@ -373,35 +381,43 @@ fi
...
 elif [ $machname == "aims4" ]; then
-  module use /usr/local/e3sm_unified/modulefiles
-  module load e3sm-unified/1.1.3-py2-nox
+  source /usr/local/e3sm_unified/envs/base/etc/profile.d/conda.sh
+  conda activate e3sm_unified_1.2.0_py2.7_nox
   export NCO_PATH_OVERRIDE=No
 elif [ $machname == "lanl" ]; then
   module unload python
   module use $projdir/modulefiles/all
-  module load e3sm-unified/1.1.3
```
Comment (milenaveneziani) [inline run_aprime.bash]: need to verify appropriate change for LANL machines

**Comment by sterlingbaldwin** [comment]
Although there's a bunch of model data on drives attached to acme1/aims4 I dont know of any thats in the directory format that aprime expects it to be in, for example all the atm/ocn/lnd data from the piControl run is there, but its in the directory structure that esgf wants, and is missing things like the meridianalHeatTransport files and restart files. 

I have some small test data available that can be used as the input directory for aprime, anyone in the "climate" group should be able to access it here: /p/user_pub/e3sm/baldwin32/testing2/many_model/output/temp/1pctCO2/aprime/0001_0002_vs_obs/20180215.DECKv1b_1pctCO2.ne30_oEC.edison/run

and the run script here: /p/user_pub/e3sm/baldwin32/testing2/many_model/output/scripts/aprime_0001_0002_1pctCO2_vs_obs.bash

If you're going to run it, please change the output directory.

**Comment by milenaveneziani** [comment]
Thanks @sterlingbaldwin for pointing us to that.

**Comment by xylar** [comment]
@milenaveneziani and @kevans32, I don't have write permission on `/space2/diagnostics` on `acme1`.  This folder belongs to @kevans32 and its subfolders belong to a user `branstetter2`.

For now, I'm going to test downloading somewhere else but we will need to have this folder opened up or deleted.

**Comment by milenaveneziani** [comment]
I didn't even know that one existed.. @kevans32: could you please open that folder up or rename it, if possible? In this PR I have included a `download_analysis_data` script that gets all the data needed by a-prime from the public html server to a certain directory. As a default, I have assumed that we can have everything in $projdir/diagnostics, but again I didn't know that this directory already exists on acme1/aims4 (it doesn't everywhere else where we support a-prime).

**Comment by kevans32** [comment]
hi guys- I am not at work today, so I have not followed this whole thread to know what you need beyond opening a directory. To keep you guys working, I "chmod 777 diagnostics". Everything below it is owned by @mlbranstetter, so Marcia can you open up everything within?

**Comment by kevans32** [comment]
If you need more from me- I can do more in the AM when I am near my computer.

**Comment by milenaveneziani** [comment]
I think this is fine for now. The data can be put under diagnostics/

**Comment by kevans32** [comment]
Wait, so reading more above.... would you prefer I rename it? ACME has a different directory structure (or at least it did) so we use a different naming protocol here.

**Comment by milenaveneziani** [comment]
if you can rename it without affecting other things, that would be good. It's up to you.
Otherwise the group has to have write permission, so that Xylar can transfer over the data needed for a-prime.

**Comment by xylar** [comment]
@kevans32, I agree with @milenaveneziani that a rename would be preferable.  There's an observations subdirectory that I would need to write to but don't have permission.

**Comment by milenaveneziani** [comment]
anyways, I think testing can be done using say `diagnostics_new`, and we can worry about renaming of directories after the holidays.

**Comment by xylar** [comment]
This is what I'm doing for now.  There's a `diagnostics_temp` where I just downloaded the data.

**Comment by sterlingbaldwin** [comment]
@xylar Sorry the data disappeared! I should have copied everything out and pointed you to something outside my test directory, those links get removed and regenerated every time I run and didnt think of that before pointing you to it.

**Comment by xylar** [comment]
@sterlingbaldwin, no worries.  I'm in the process of downloading 3 years of a run myself for this purpose so nothing you need to worry about at this point.

**Comment by milenaveneziani** [comment]
While testing on theta, using the login node (compute nodes can't be used because they are for large jobs only, and we only need one single node), I was told to stop because a-prime was using way too much memory. The solution is to run on cooley instead.
After @xylar installs a new conda environment for cooley, I will test again there, and change the setting of `run_aprime` so that it points to cooley and not theta on ALCF.

**Comment by milenaveneziani** [comment]
Testing on cooley's login node is successful with new conda environment.
Now I need to add some working batch scripts for cooley.

**Comment by milenaveneziani** [comment]
Testing on titan's login node works, but the batch job fails for MPAS-Analysis for reasons that are not clear to me (climatology tasks just hang without producing anything, and the streamfunction MOC doesn't even start. No error message is produced, the tasks just seem to vanish).

**Comment by rljacob** [comment]
@milenaveneziani and @xylar Note that theta project directories are now visible from cooley. This is very recent.  You should be able to see /lus/theta-fs0/projects/ClimateEnergy_2 from cooley.

**Comment by milenaveneziani** [comment]
Hi @rljacob: yes, they are visible, and that is great, but the problem is that we cannot run big jobs (analysis on high-res data) on cooley's login node (or theta's), so we need to run batch jobs, and I have been getting errors because ClimateEnergy_2 is not associated with cooley.. Which does not make much sense to me: if we have to use cooley for analysis, then we need to be able to run on its compute nodes. So, I think that, instead of being an automatic thing, we now need to request a cooley allocation for ClimateEnergy_2.

**Comment by kevans32** [comment]
@milenaveneziani so I can jump in and assist with testing, I pulled the latest to try on titan and got this error:
IOError: [Errno 2] No such file or directory: u'/lustre/atlas1/cli115/proj-shared/ACME_simulations/20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison/run/mpas-o_in'
If you are mid testing, I will wait, but just thought I would see if I see same as you. I will have more time when I return to work on Monday, but FYI in the 
/lustre/atlas1/cli115/proj-shared/ACME_simulations 
directory all I see is this directory:
 20180410.A_WCYCL1950_HR.ne120_oRRS18v3_ICG.theta,
which is interesting.. more to come on Monday.

**Comment by milenaveneziani** [comment]
Hi @kevans32: thanks for jumping in. That directory was purged, so there is no model data in there. I removed the whole thing yesterday. I now have 3 years of high-res on titan. That will probably be too much to run on the login node right now (and my test of the batch script did not complete on titan). So, I'd say hold off testing at OLCF now, until we have some low-res data available there. No E3SM run is being run at OLCF at the moment, as far as I know.

If you have some time and access to anvil, you could try there.

**Comment by kevans32** [comment]
Sorry I don't have anvil access. If you point me to some good example data at NERSC, I can copy it over for testing.

**Comment by milenaveneziani** [comment]
ah, you can also test on nersc, forgot about that.

A good example from NERSC is the piControl from the deck experiment:
/global/cscratch1/sd/golaz/ACME_simulations/20180129.DECKv1a_piControl.ne30_oEC.edison/archive
I tried to transfer 3-5 years of this over to OLCF with Globus, but I kept getting connection errors with NERSC. Not sure if today is better. The location to put it would be:
/lustre/atlas1/cli115/proj-shared/ACME_simulations
You should have permission to make a new directory in there (20180129.DECKv1a_piControl.ne30_oEC.edison/archive/etcetc) and copy stuff.

**Comment by milenaveneziani** [comment]
Batch job script finally works on cooley, so we should be set for ALCF. Last testing to do is on acme1/aims4. I also would like to give it a go on the LANL machines, because something tells me that the current setting will no longer work there.
Last thing to do will be to update to the official MPAS-Analysis v1, hopefully soon.

@kevans32: is there anything else you would like to add to this release?

**Comment by xylar** [comment]
@milenaveneziani, I will try to test on acme1 and aims4 today.

File: bash_scripts/batch_ocnice.cooley.bash [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+#COBALT -t 2:00:00
+#COBALT -n 1
+#COBALT -A OceanClimate_2
```
[the remark this answers is not in the document]
Comment (milenaveneziani) [reply bash_scripts/batch_ocnice.cooley.bash]: yes, I'll switch in the final rebase.

[the remark this answers is not in the document]
Comment (xylar) [reply bash_scripts/batch_ocnice.cooley.bash]: Great!

**Comment by milenaveneziani** [comment]
@xylar: could you double check run_aprime, especially for the part that deals with proj directories and with the acme1/aims4 setup?
The only thing left to do at this point, besides a quick rebase, is the MPAS-Analysis update to the official v1 tag.

**Comment by xylar** [comment]
@milenaveneziani, I'm testing on acme1 right now.  So far, it's still on the atmosphere part but all seems to be going well.  I'll keep you posted.

**Comment by xylar** [comment]
@milenaveneziani, is it expected behavior that `parallelTaskCount = 1` if you run on a login node even if I set `mpas_analysis_tasks = 4`?  Similarly, `ncclimoParallelMode` is ignored.  I assume this is a safety measure but it seems like this means I can't run in parallel on acme1 (which has no batch queue that I'm aware of).  The atmosphere diagnostics seem to have no such safety measure and run dozens of python tasks at once.

**Comment by milenaveneziani** [comment]
yes, I have only dealt with parallel tasks on compute nodes/when using batch scripts. How would you suggest we handle this instead?

**Comment by xylar** [comment]
I guess I would trust the user to set a reasonable value for `mpas_analysis_tasks` and the corresponding `ncclimo` flag, and use it both on the login node and in batch mode.  Or at least do this on select machines where batch mode is not available.

**Comment by xylar** [comment]
It looks like the `ncclimo` flag is set regardless of batch mode, so I was wrong about that one.

**Comment by milenaveneziani** [comment]
yes, I think it's the `parallel_task_count` that is only set by the batch script, if using it. The ncclimo flag is set to 12 by default.
One possibility would be to set parallel_task_count=mpas_analysis_tasks only on some machines, like acme1.

**Comment by milenaveneziani** [comment]
I have decided to leave the treatment of `parallel_task_count` as is for now, since it seems a bit cumbersome to me to have it defined only for one machine. I could also define it all the times, but then is it OK to use more than one processor on a login node, were the user to forget to set `mpas_analysis_tasks` to 1?

I have updated to the official MPAS-Analysis v1 tag, and rebased. Let me know how the test on acme1 goes.

**Comment by xylar** [comment]
Wouldn't it then make more sense to have similar protections for the atmosphere diagnostics and the ncclimo calls, which are both much more computationally intensive than, say 4-6 parallel tasks?

**Comment by xylar** [comment]
Part of it is that acme1 is intended as a dedicated analysis machine for e3sm so I do think it's worth the trouble of treating it as a special case.

My suggestion would be to set mpas_analysis_tasks to 1 if a user is running on a machine other than acme1 or aims4 and not in batch mode. I also think a warning should be displayed that this has been done. Then, the number of parallel tasks will always be mpas_analysis_tasks.

I still think this safeguard combined with ncclimoParallelMode = bck and the massive number of threads generated by the atmosphere diagnostics is a little inconsistent. Basically, a-prime should be run in batch mode on all machines other than acme1 to keep us from getting in trouble on the login nodes.

**Comment by milenaveneziani** [comment]
> a-prime should be run in batch mode on all machines other than acme1 to keep us from getting in trouble on the login nodes.

We never got in trouble with low-res model data. But definitely the high-res cannot be run on the login node, I agree.

I can make the change for mpas_analysis_tasks.

What happened with the acme1 job, did it finish successfully?

**Comment by xylar** [comment]
Yes, it did. Everything looked like it ran fine. I didn't get a chance to look at the web page yet but I don't expect any surprises.

**Comment by milenaveneziani** [comment]
Everything worked on acme1 for me too.
@xylar: let me know if you think anything else should be changed.

**Comment by milenaveneziani** [comment]
@sterlingbaldwin: would you like to test this on acme1 or aims4 to make sure that you don't have the problems you were having before?

**Comment by xylar** [comment]
The changes look great, @milenaveneziani. Given how you implemented it, I don't think the warning is necessary so you can take it out if you want. It will be clear how someone could edit the file to get a different behavior and also when to expect 1 vs 12 tasks.

**Comment by milenaveneziani** [comment]
> I don't think the warning is necessary so you can take it out if you want.

ok, I took the warning out and added some more info in the run_aprime documentation.

Would you mind formally approving this and I think I will merge and create a v1.2 tag, which would be the mini-release that @kevans32 and I talked about (@kevans32: do you agree?).
@sterlingbaldwin: you can always test this on acme1/aims4 when you have some time and open an issue if you encounter any problems.

**Comment by kevans32** [comment]
Thanks for all these efforts! I do want it working on OLCF systems too- can you point me to data to transfer over and try to use it there? Or are there changes that would have broken the OLCF version? I hope not.

**Comment by milenaveneziani** [comment]
no no, everything works at OLCF. I tested there with high-resolution model output (for which you can only use the run_batch_script=true option).
We could try again transferring low-res data from nersc to olcf, so that you can test with that. Last time around, I had trouble with globus on nersc, for some reason.

**Comment by kevans32** [comment]
Also before you make the tag and release we need to make sure zenodo will pick it up and assign it a DOI. I looked into this some weeks ago with help from @rljacob . So let me look into that again and make sure its ready before you click go.

**Comment by milenaveneziani** [comment]
Would you have to check that with the actual tag, or this branch is fine?

**Comment by kevans32** [comment]
hold on- computer died! going away for a bit....

**Comment by milenaveneziani** [comment]
I have initiated a globus transfer to have a low-res run example at OLCF. The data will be here:
/lustre/atlas1/cli115/proj-shared/ACME_simulations/20180129.DECKv1b_piControl.ne30_oEC.edison

Once the transfer is done, would you like to run a test with it, @kevans32? You will need to update your run_aprime script, and could use my example here:
`~milena/a-prime/run_aprime_rhea.bash`

**Comment by milenaveneziani** [comment]
@kevans32: data transfer is done and you could start a test case at OLCF, if you want.

**Comment by milenaveneziani** [comment]
you also wanted to check the zenodo thing, right?
We can do this tomorrow morning.

**Comment by kevans32** [comment]
Regarding the DOI, a-prime appears to be set up to notify zenodo when we publish a release: https://github.com/E3SM-Project/a-prime/settings/hooks. I can see from my zenodo account that a-prime is an enabled repostory ( I cannot change it, but its on thanks to @rljacob), and so when you publish, a tarball of it should be automatically uploaded and DOI created. Make sure when you publish its the version you want to go public. So zenodo is ready when you are!

**Comment by milenaveneziani** [comment]
ok, that seems consistent with what @xylar was telling me this morning.

What does 'publish' mean in this sense? When I make a new tag, or do I have to do a release on github? or?

**Comment by kevans32** [comment]
Here is a nice page that lays it out: https://help.github.com/articles/creating-releases/

**Comment by xylar** [comment]
> What does 'publish' mean in this sense? When I make a new tag, or do I have to do a release on github? or?

I believe this is triggered when you create a release by clicking on the "Draft a new release" button.  I'm not sure what the equivalent on the command line is.  But creating a tag on the command line is probably not enough.  I think tags aren't always releases in the sense of creating a tarball of the code, etc.

**Comment by rljacob** [comment]
I'm pretty sure it means when you do a release on github: pushing the "Draft a new release" button on  https://github.com/E3SM-Project/a-prime/releases.  Confusingly, github will label every tag as a "release" but its really something they call a "pre-release".  (we went around with them on this before E3SM v1.0.0)

**Comment by kevans32** [comment]
Then, there is one more button, "publish release," that you click to send in order to for zenodo to pick it up.

**Comment by milenaveneziani** [comment]
ok, great, thanks @rljacob. I'll create a tag, push it and then turn to github to do the release thing.

@kevans32: let me know how your test goes tomorrow, and I will get this done. Thanks.

**Comment by xylar** [comment]
>  Confusingly, github will label every tag as a "release" but its really something they call a "pre-release". 

If that was true before, I think it's no longer the case, having created a release myself last night.  There's a check box for "pre-release" but the default it's not checked, meaning a normal release.

**Comment by kevans32** [comment]
Tests on titan and rhea with low and high res were successful ( of course you made it trivial for me by providing the scripts with the data path loaded- thanks!). @milenaveneziani  @xylar I am ready when you are ready.

**Comment by milenaveneziani** [comment]
Great, thanks @kevans32 for testing!

**Pull request merged**